jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/CombinedResource.java

```diff
@@ -252,24 +260,37 @@ public Iterator<Resource> iterator()
     @Override
     public List<Resource> list()
     {
-        List<Resource> result = new ArrayList<>();
-        for (Resource r : _resources)
+        Map<String, Resource> results = new TreeMap<>();
+        for (Resource base : _resources)
         {
-            if (r.isDirectory())
-                result.addAll(r.list());
-            else
-                result.add(r);
+            for (Resource r : base.list())
+            {
+                if (r.isDirectory())
+                    results.computeIfAbsent(r.getFileName(), this::resolve);
+                else
+                    results.putIfAbsent(r.getFileName(), r);
+            }
         }
-        return result;
+        return new ArrayList<>(results.values());
     }
 
     @Override
     public void copyTo(Path destination) throws IOException
     {
-        // Copy in reverse order
-        for (int r = _resources.size(); r-- > 0; )
+        Collection<Resource> all = getAllResources();
+        for (Resource r : all)
         {
-            _resources.get(r).copyTo(destination);
+            Path pathTo = getPathTo(r);
+            Path to = destination.resolve(pathTo);
+            if (r.isDirectory())
+            {
+                if (!Files.exists(to))
+                    Files.createDirectories(to);
+            }
+            else
+            {
+                Files.copy(r.getPath(), to, StandardCopyOption.COPY_ATTRIBUTES, StandardCopyOption.REPLACE_EXISTING);
+            }
```

Comment:
This implementation is broken.

This shouldn't be doing the Path manipulation here.
The use of `getAllResources()` is wrong here, it should be delegating to each implementation's `Resource.copyTo(destination)` here instead.
That way cross-filesystem type copyTo is supported.

I have a testcase that breaks this implementation, and a fix in a branch locally I will be happy to demonstrate.

Reply:
@joakime The original code did do a delegation to each nested Resource in reverse order.   That is an algorithm that results in the correct result, but is very inefficient if large resources are overridden, since they will be copied and then replaced.
This algorithm avoid that double copy and the use of getAllResources is exactly the semantic we want to achieve.

But as you say,   we may need to duplicate the cross file system copy here.....      So either that code get's factored out to be usable by both impls of `copyTo` or we revert to the previous algorithm and live with the possible inefficiencies.

Reply:
I've fixed it with a cross FileSystem resolve trick.... I'll move that to a util somewhere